src/ophyd_async/epics/pvi/pvi.py

```diff
@@ -0,0 +1,280 @@
+            for sub_sub_entry in entry.sub_entries[sub_name].values():  # type: ignore
+                _verify_common_blocks(sub_sub_entry, sub_device)  # type: ignore
+        else:
+            _verify_common_blocks(
+                entry.sub_entries[sub_name], sub_device  # type: ignore
+            )
+
+
+_pvi_mapping: Dict[FrozenSet[str], Callable[..., Signal]] = {
+    frozenset({"r", "w"}): lambda dtype, read_pv, write_pv: epics_signal_rw(
+        dtype, "pva://" + read_pv, "pva://" + write_pv
+    ),
+    frozenset({"rw"}): lambda dtype, read_write_pv: epics_signal_rw(
+        dtype, "pva://" + read_write_pv, write_pv="pva://" + read_write_pv
+    ),
+    frozenset({"r"}): lambda dtype, read_pv: epics_signal_r(dtype, "pva://" + read_pv),
+    frozenset({"w"}): lambda dtype, write_pv: epics_signal_w(
+        dtype, "pva://" + write_pv
+    ),
+    frozenset({"x"}): lambda _, write_pv: epics_signal_x("pva://" + write_pv),
+}
+
+
+def _parse_type(
+    is_pvi_table: bool,
+    number_suffix: Optional[int],
+    common_device_type: Optional[Type[Device]],
+):
+    if common_device_type:
+        # pre-defined type
+        device_type = _strip_union(common_device_type)
+        is_device_vector, device_type = _strip_device_vector(device_type)
+        import inspect
+
+        if ((origin := get_origin(device_type)) and issubclass(origin, Signal)) or (
+            inspect.isclass(device_type) and issubclass(device_type, Signal)
+        ):
+            # if device_type is of the form `Signal` or `Signal[type]`
+            is_signal = True
+            signal_dtype = get_args(device_type)[0]
+        else:
+            is_signal = False
+            signal_dtype = None
+
+    elif is_pvi_table:
+        # is a block, we can make it a DeviceVector if it ends ina number
+        is_device_vector = number_suffix is not None
+        is_signal = False
+        signal_dtype = None
+        device_type = Device
+    else:
+        # is a signal, signals aren't stored in DeviceVectors unless
+        # they're defined as such in the common_device_type
+        is_device_vector = False
+        is_signal = True
+        signal_dtype = None
+        device_type = Signal
+
+    return is_device_vector, is_signal, signal_dtype, device_type
+
+
+def _sim_common_blocks(device: Device, stripped_type: Optional[Type] = None):
+    device_t = stripped_type or type(device)
+    for sub_name, sub_device_t in get_type_hints(device_t).items():
+        if sub_name in ("_name", "parent"):
+            continue
+
+        # we'll take the first type in the union which isn't NoneType
+        sub_device_t = _strip_union(sub_device_t)
+        is_device_vector, sub_device_t = _strip_device_vector(sub_device_t)
+        is_signal = (
+            (origin := get_origin(sub_device_t)) and issubclass(origin, Signal)
+        ) or (issubclass(sub_device_t, Signal))
+
+        if is_device_vector and is_signal:
```

Comment:
Does the device_vector have a pre-determined size, or is this returning a size 2 always ok?

Reply:
In the non-sim mode of the panda, there can be up to four `SEQ` blocks, and a variable number for other blocks. There's no way to know how many prior to introspection. In sim mode I've just had the device be set up with everything in the common class, and if there's a device vector I've arbitrarily decided on two elements.